src/main/java/net/logstash/logback/appender/AbstractBeatsTcpSocketAppender.java

```diff
@@ -0,0 +1,373 @@
+    private static final byte PAYLOAD_WINDOW_TYPE = 'W';
+    private static final byte PAYLOAD_ACK_TYPE = 'A';
+
+    /**
+     * Counter for sequence number of published events.
+     * Sent to remote Beats input to track how many messages were sent
+     * and used to track when to wait for remote to be ready to accept new logs.
+     */
+    private final AtomicInteger counter = new AtomicInteger();
+    /**
+     * A queue of currently accepted ACKs from remote
+     */
+    private final BlockingQueue<AckEvent> ackEvents = new ArrayBlockingQueue<>(10);
+
+    /**
+     * Used to tell the reader the maximum number of unacknowledged data frames
+     * the writer (this appender) will send before blocking for ACKs.
+     */
+    private int windowSize;
+
+    /**
+     * Callable that reads ACKs from remote reader.
+     */
+    private AckReaderCallable ackReaderCallable;
+
+    public AbstractBeatsTcpSocketAppender() {
+        this.windowSize = 10;
+    }
+
+    @Override
+    public synchronized void start() {
+        Encoder<Event> encoder = getEncoder();
+        setEncoder(wrapAsBeatsEncoder(encoder));
+        super.start();
+    }
+
+    private <E> Encoder<E> wrapAsBeatsEncoder(Encoder<E> original) {
+        return new BeatsEncoderWrapper<>(original, createConverter());
+    }
+
+    private LumberjackV2PayloadWrapper createConverter() {
+        return new LumberjackV2PayloadWrapper(
+                counter, windowSize
+        );
+    }
+
+    public void setWindowSize(int windowSize) {
+        this.windowSize = windowSize;
+    }
+
+    public int getWindowSize() {
+        return windowSize;
+    }
+
+    /**
+     * Beats-compatible encoder.
+     * Wraps the original encoder and blocks if we should wait for ACK.
+     *
+     * @param <E> log event type
+     */
+    private class BeatsEncoderWrapper<E> extends EncoderBase<E> {
+
+        private final Encoder<E> original;
+        private final LumberjackV2PayloadWrapper wrapper;
+
+        BeatsEncoderWrapper(Encoder<E> original, LumberjackV2PayloadWrapper wrapper) {
+            this.original = original;
+            this.wrapper = wrapper;
+        }
+
+        @Override
+        public boolean isStarted() {
+            return original.isStarted() && super.isStarted();
+        }
+
+        @Override
+        public void start() {
+            if (isStarted()) {
+                return;
+            }
+            original.start();
+            super.start();
+        }
+
+        @Override
+        public void stop() {
+            if (isStarted()) {
+                super.stop();
+                original.stop();
+            }
+        }
+
+        @Override
+        public byte[] headerBytes() {
+            return original.headerBytes();
+        }
+
+        @Override
+        public byte[] encode(E event) {
+            if (counter.get() == windowSize) {
+                try {
+                    ackEvents.take();
```

Comment:
The `WriteTimeoutRunnable` may decide to close the socket whenever it detects a _dead peer_ (that is a write operation taking _too long_). At this time the EventHandler thread may be waiting for the next ACKs before writing the current event. If the socket is closed, the ACK will never come and the thread will not exit from `ackEvents.take()` (the AckReaderCallable will "die" because of the IOException received when the socket is closed).